### PR TITLE
Introduce beresp.fail_reason & beresp.fail_detail

### PR DESCRIPTION
**Pull request by stevendore**
These are two variables only visible in vcl_backend_error. These variables describe the reasoning for reaching vcl_backend_error.

`beresp.fail_reason`  will always have a value. The value should be picked from the pre-defined list found in tbl/backend_fail.h. This should not contain sensitive information (hostnames, IP address, etc.).

`beresp.fail_detail` like `beresp.fail_reason` but contains additional or custom information. It may contain sensitive information. It will not be put in client response by ``builtin.vcl``. Suggested format: `<director> : <details>`.

This was proposed during [VDC](https://etherpad.wikimedia.org/p/vdd19q3).

Some notes on this PR:

1) Using table to create string variables isn't very clean. Maybe breaking off from table to a single header file of macros would be cleaner.
2) How many VTCs should be included? One for each error? a handful?
3) A VTC using VMOD Debug to make custom error messages should be added as well

### CODE REVIEW

File: bin/varnishd/cache/cache_backend.c [hunk 1]
```diff
@@ -125,9 +126,12 @@ vbe_dir_getfd(struct worker *wrk, struct backend *bp, struct busyobj *bo,
 	CHECK_OBJ_NOTNULL(bp, BACKEND_MAGIC);
 	AN(bp->vsc);
 
+	bo->fail_reason = BE_FAIL_UNKNOWN;
```
Comment (nigoroll) [inline bin/varnishd/cache/cache_backend.c]: is this really the best place for the initialization? I wonder if it should rather be moved to when we create the busyobj

File: bin/varnishd/Makefile.am [hunk 2]
```diff
@@ -150,6 +151,7 @@ noinst_HEADERS = \
 nobase_pkginclude_HEADERS = \
 	cache/cache.h \
 	cache/cache_backend.h \
+	cache/cache_backend_fail.c \
```
Comment (nigoroll) [inline bin/varnishd/Makefile.am]: .h rather?

**Comment by fgsch** [comment]
I've been a long proposer for something like this so 👍 

Personally I'd prefer if we expose this in a separate namespace, e.g. `error`, and add attributes, such as `error.reason` and `error.detail` so it could potentially be reused in other subroutines, and vmods as well.

**Comment by stevendore** [comment]
> Personally I'd prefer if we expose this in a separate namespace, e.g. `error`, and add attributes, such as `error.reason` and `error.detail` so it could potentially be reused in other subroutines, and vmods as well.

This sounds interesting. I think it is above my pay-grade so to speak. When was the last time a new namespace has been added?

File: bin/varnishd/cache/cache_backend.c (hunk 1 above)
Comment (stevendore) [reply bin/varnishd/cache/cache_backend.c]: Makes sense. Thanks for bringing it up. I have moved it.

**Comment by stevendore** [comment]
> I am mssing the error vmod we talked about (https://etherpad.wikimedia.org/p/vdd19q3)

Yup working on that now. Do we have a way to tbl-ify the .vcc file like we do in C files? I don't think so. Otherwise for now I am going to do it by hand.

**Comment by nigoroll** [comment]
> > Personally I'd prefer if we expose this in a separate namespace, e.g. `error`, and add attributes, such as `error.reason` and `error.detail` so it could potentially be reused in other subroutines, and vmods as well.
> 
> This sounds interesting. I think it is above my pay-grade so to speak. When was the last time a new namespace has been added?

A vmod is your namespace.

**Comment by nigoroll** [comment]
> > I am mssing the error vmod we talked about (https://etherpad.wikimedia.org/p/vdd19q3)
> 
> Yup working on that now. Do we have a way to tbl-ify the .vcc file like we do in C files?

I am still missing a nice way, but I do (ab)use cpp to do this here: https://code.uplex.de/uplex-varnish/varnish-objvar/blob/master/src/Makefile.am#L68

> Besides do we have an API in mind? something like `error.backend_fail(ENUM)`?

I think it would rather be just one function per error, like `$Function STRING be_fail_sick()`, just returning the fixed string, so you could say in VCL

```
import error;
sub vcl_backend_error {
    if (beresp.fail_reason == error.be_fail_sick()) {
        ...
    }
}
```

As mentioned during the VDD, we could quite easily introduce vmod constants, so the above would become something like `$Constant STRING be_fail_sick`

```
import error;
sub vcl_backend_error {
    if (beresp.fail_reason == error.be_fail_sick) {
        ...
    }
}
```

later

**Comment by nigoroll** [comment]
@bsdphk you once mentioned that the constants in VCC would be a no-brainer, would you maybe just add them?

**Comment by nigoroll** [comment]
force-pushed to rebase onto current master

**Comment by nigoroll** [comment]
@Dridi on your _[1]_ it's not even my PR ;)

**Comment by stevendore** [comment]
> do we need a new `cache_backend_fail.[ch]`?
> * the constants can be moved to `cache_backend.[ch]`

That works. I went with an overly complicated solution when the easy one was right there. Thanks.

> inconsistent naming
> RX_JUNK vs OVERFLOW
> FTBO vs CONNECT_TO (I'd prefer FIRSTBYTE and CONNECT)

Naming things is always hard but the vocabulary (variable name and strings) could have been done better. I will spend more time on that and use your suggestion.

> maybe not UNKNOWN as a catch-all, I'd prefer UNSPEC
> in any case, we agree that this one is for third-party director errors that don't match a specified one?

> in some (all?) cases the fail reason is part of the error log, but is duplicated in the code instead of reusing the fail_reason field

I was avoiding non-additive code but I do agree this should be tied together. Places with errnos will need something slightly different.



UNSPEC is much better.

> in makefile and include tables lexicographic order is broken (in some cases it was already broken)

Them being broken in a lot of places threw me off but I will see if I can do better to make it fit where needed (Though I think the above comments remove the need for this one)


> In general I really don't want to commit to these new symbol names and would rather take the time to go over them. Maybe we should expose them in VCL as #3134 constants? Just kidding... (Or am I?)

One of the main goals of this PR, for me, was working on improving the vocabulary

**Comment by stevendore** [comment]
> in some (all?) cases the fail reason is part of the error log, but is duplicated in the code instead of reusing the fail_reason field

I took another look at this and the VSL errors and the name of the errors don't always align well. It then felt odd to change some but not others but maybe I am wrong there. For example: 

Backend busy contains the name of the backend in it. This is similar for backend sick.
https://github.com/varnishcache/varnish-cache/blob/3f4ff9e1858bbe13b94aa1a1bc9e96efb7cbbd09/bin/varnishd/cache/cache_backend.c#L141-L143

Or first byte timeout isn't clear which timeout it was but that could be implicit in the VSL seeing a connection but I am not sure off hand:

https://github.com/varnishcache/varnish-cache/blob/3f4ff9e1858bbe13b94aa1a1bc9e96efb7cbbd09/bin/varnishd/cache/cache_backend.c#L282-L288

**Comment by hermunn** [comment]
Is everything ready here? @stevendore, are you waiting for a final review / OK?

**Comment by stevendore** [comment]
Waiting for final review.

**Comment by nigoroll** [comment]
> My main problem otherwise is vmod_error

This had actually found agreement at the last VDD

**Comment by nigoroll** [comment]
@stevendore can you please rebase?

File: bin/varnishtest/vmods.h [hunk 3]
```diff
@@ -28,8 +28,8 @@
  */
 
 VTC_VMOD(std)
-VTC_VMOD(error)
 VTC_VMOD(debug)
+VTC_VMOD(error)
 VTC_VMOD(directors)
```
[the remark this answers is not in the document]
Comment (stevendore) [reply bin/varnishtest/vmods.h]: fixed by this file not existing anymore :) but other places where alphabetic order was out of place has been fixed as well.

File: bin/varnishd/cache/cache_backend.c [hunk 4]
```diff
@@ -60,6 +59,9 @@ static VTAILQ_HEAD(, backend) cool_backends =
     VTAILQ_HEAD_INITIALIZER(cool_backends);
 static struct lock backends_mtx;
 
+#define BACKEND_FAIL(U,l, s)	const char *BE_FAIL_##U = s;
```
[the remark this answers is not in the document]
Comment (stevendore) [reply bin/varnishd/cache/cache_backend.c]: Thanks! fixed.

File: bin/varnishd/cache/cache_backend.h [hunk 5]
```diff
@@ -72,6 +72,9 @@ struct backend {
 	vtim_real		cooled;
 };
 
+#define BACKEND_FAIL(U,l, s)	extern const char *BE_FAIL_##U;
```
[the remark this answers is not in the document]
Comment (stevendore) [reply bin/varnishd/cache/cache_backend.h]: Thanks! fixed.

**Comment by stevendore** [comment]
> @stevendore can you please rebase?

This has been rebased on master. I have not squashed it yet in case someone wanted to look through the commits but let me know when squashing is wanted.

**Comment by bsdphk** [comment]
This was discussed extensively between @nigoroll, @Dridi and @bsdphk last week, and we found that 300 added LOC was a LOT for the gains.

We came up instead with the following more general concept:

1.  We need a `{be}resp.whyamihere` kind of `STRING` variable in `vcl_synth{}` and `vcl_backend_error{}`.

Precise naming to be bikeshedded, as the obvious `{be}resp.reason` is already taken.
Candidate names could be (`.cause`, `.why`, ...)

2. We need named $ENUM's in vmods.

So that one can say `if (beresp.whyamihere == foovmod.enumname.enumvalue) {...}`

3. We can consider introducing `is` as an VCL expr operator.
The idea is that it checks pointer identity, rather than strcmp.
Not really obvious if this is actually worth it, or if it will only have people make new kinds of mistakes.

`if (beresp.whyamihere is foovmod.enumname.enumvalue) {...}`

4. We need an enum for the built-in backen errors.  Current namespace would allow us to use `backend.error.noconnection` etc.

**Pull request closed without merging**